### DISCUSSION TITLE
"Request timeout. Please try again later."

### DISCUSSION THREAD

**Discussion by retry-the-user** [deepgram/community#367] [General help]
### Which Deepgram product are you using?

Deepgram API

### Details

Occasionally when we submit audio for STT to the API, this is the final response to our webhook endpoint.  Keep in mind this is what you refer to as the pre-recorded processing where we send you audio in step one, then you respond to us with the answer in step two.  We are not blocking and waiting for a response during the initial send.

The first part of this is successful.  We send you the audio; you respond with success and a unique identifier.  If the error is in the audio file sending, then we'd expect to see the error in step 1.  This would lead to our queue system retrying later.  

The problem is you say you successfully received our audio, then later respond that you timed out.  If your STT process internally timed out, why tell us? Why not retry and send us the proper response?  This feels like you're passing the buck on to end users for a retrying mechanism on what looks like an internal problem with your job process.

Example (sent to our URL endpoint by Deepgram):
{"err_code":"Gateway Timeout","err_msg":"Request timeout. Please try again later.","request_id":"a636d084-473d-431c-98c4-8a3cb63f787a"} (received at 2023-09-26 17:53 -- 10 minutes later)

And here is our log of when we sent you that file:
IcQ9RPvt.wav    a636d084-473d-431c-98c4-8a3cb63f787a    size: 1012350   stt_req: 0.64   2023-09-26 17:43:48

A 1MB wav file was sent in .64 seconds, and you responded with that UUID.  So if you possess the file, why wouldn't you process it until you succeed before sending us a response? What good does this error message do us at this point? Either fail on the initial send or retry STT until you succeed and send us an actual answer.

As far as I'm concerned, this is a bug or service failure.  Either way, if you can't handle processing, then throw the error early and reject new incoming files until your processing can handle the load again.

### If you are making a request to the Deepgram API and have a request ID, please paste it below:

a636d084-473d-431c-98c4-8a3cb63f787a

### If possible, please attach your code or paste it into the text box.

Intermittent issue.  There is nothing wrong with the code which works under normal conditions.

### If possible, please attach an example audio file to reproduce the issue.

Intermittent issue. Nothing to do with the specific audio.

**Comment by retry-the-user** [comment]
Today for the first time we're getting a new intermittent error: "SLOW_UPLOAD"

Example: 
!----------2023-10-11 19:30:27----------!
{"err_code":"SLOW_UPLOAD","err_msg":"Request upload timeout.","request_id":"65d000a7-670a-46f3-befa-bbb094125292"}

That's the JSON response from DG to our webhook URL.  The initial upload logged like this:

IzdSA7gE.wav    65d000a7-670a-46f3-befa-bbb094125292    size: 109772    stt_req: 7.11   2023-10-11 19:30:21

So it looks like it took 7 seconds to upload the 109KB file, but then there was this error response 6 seconds later.  Whose fault that the upload was slow? Why do I get this error on the back end? Where's the STT result?  This is just horrendous error handling on your part.

**Comment by lukeocodes** [comment]
Hi @retry-the-user. Can you confirm if you're using express, or a serverless function as the webhook?

**Reply by retry-the-user** [reply]
I apologize for not noticing your  question earlier. I don't understand what you're asking.  The webhook is a URL to an httpd server. Does that mean express?  This problem has not gone away and continues to plague us:

Sending wav file:
3984430.wav    86117507-42a1-4c8e-a4ae-c84110c235ed    size: 68550     stt_req: 30.35  2023-11-04 16:10:36    Deepgram

I don't know why it took 30 seconds to send you a 68K file.  We're not bottle-necked. In fact our site is close to idle on Saturdays. These rate limits are on your end.

!----------2023-11-04 16:10:42----------!
{"err_code":"SLOW_UPLOAD","err_msg":"Request upload timeout.","request_id":"685ca88a-7d73-4755-adaa-545d9f4802b5"}

This confuses me.  The response came back before the file transfer even completed AND the request id is different. But this is the only match to that wave file.

**Reply by retry-the-user** [reply]
I'll have to do some deeper analysis of our log files, but from what I can see, sometimes these slow request errors come in and yet later you did in fact process the audio and return a text result. But sometimes not. It's inconsistent, and that's even more troublesome. 

If I'm going to retool our procedure to retry to send the audio after we've gotten a success message on the upload side when you send these out-of-sync error messages, then we will end up needlessly resubmitting audio to you, which will just further create load on your servers that maybe you're already struggling with (hence the slow uploads in the first place).

**Reply by retry-the-user** [reply]
Ok here's very peculiar example that indicates, to me at least, that this SLOW UPLOAD problem is entirely on the DG side of things.

First we sent this wav file of around 446KB in .25 seconds (hardly SLOW)
3975181    a6658d8e-a9d5-4a7a-814a-346212c39221    size: 446918    stt_req: 0.25   2023-11-03 17:44:34    Deepgram

Then our webhook received the following replies. Remember, we only uploaded the wav file to DG one time:

2023-11-03 17:41:06     ID: 3975181      Status: {"err_code":"SLOW_UPLOAD","err_msg":"Request upload timeout.","request_id":"08b14f4a-3d26-4074-900e-61926107768b"}
2023-11-03 17:41:25     ID: 3975181      Status: {"err_code":"SLOW_UPLOAD","err_msg":"Request upload timeout.","request_id":"c08e1fcb-a610-4941-81dc-2fd6bfae9bfb"}
2023-11-03 17:41:55     ID: 3975181      Status: {"err_code":"SLOW_UPLOAD","err_msg":"Request upload timeout.","request_id":"5130b049-f6a9-494b-8fdf-3d1ac1deb9d1"}
2023-11-03 17:44:35     ID: 3975181      Status:
Hi. This is Amanda from .... (message redacted)

I'm at a loss to explain why we got THREE SLOW_UPLOAD errors before the final result (all with different request id's that do not correlate to the file in question).  The timestamps are all before we even sent you the file with that ID.  I see no evidence on our end that we ever retried to send that particular wav file.  And yet we logged those errors almost 3 minutes before we sent the file with that ID.  

It looks to me like we may have gotten a failure to upload from you prior to the successful one.  But if that's true, why the SLOW UPLOAD errors?  If you had thrown us an error with an id, we'd have logged it but considered it a successful upload (we only care about the id).  If we never got one of your request_id's back, we'd have retried the request.  

Obviously I need to do a better job verifying the upload of the file was a complete success.  But these SLOW UPLOAD errors are not useful.  Sometimes the request goes through and completes (albeit late). Sometimes, we never get a result.  It's hard to build a deterministic system around such inconsistency.
